### PR TITLE
DNS ML Feedback

### PR DESCRIPTION
**Pull request by rabarona**
Adding support to DNS ML Feedback reading local file in ML node.

### CODE REVIEW

File: dns_pre_lda.scala [hunk 1]
```diff
@@ -78,51 +79,97 @@ val l_top_domains = Source.fromFile("top-1m.csv").getLines.map(line => {
...
+    16  hh                      int64
+    17  ip_sev                  int64
+    18  dns_sev                 int64
+    19  dns_qry_class_name     object
+    20  dns_qry_type_name      object
+    21  dns_qry_rcode_name     object
+    22  network_context       float64
+    23  unix_tstamp             
+    */    
+    val FrameTimeIndex = 0
```
Comment (NathanSegerlind) [inline dns_pre_lda.scala:109]: only classes and objects should be capitalized, eg. frameLenIndex and dnsQryNameIndex

also, correct camelcasing would be unixTStampIndex and honestly, why not spell it out as unixTimeStampIndex ?

File: dns_pre_lda.scala [hunk 2]
```diff
@@ -78,51 +79,97 @@ val l_top_domains = Source.fromFile("top-1m.csv").getLines.map(line => {
...
+    val UnixTstampIndex = 23
+    val FrameLenIndex = 1
+    val IpDstIndex = 2
+    val DnsQryNameIndex = 3
+    val DnsQryClassIndex = 4
+    val DnsQryTypeIndex = 5
+    val DnsQryRcodeIndex = 6
+    val DnsSevIndex = 18
+    
+    val lines = Source.fromFile(feedback_file).getLines().toArray.drop(1)
```
Comment (NathanSegerlind) [inline dns_pre_lda.scala:122]: explain the drop with a comment (I know it's a header but maybe reader doesn't)

File: dns_pre_lda.scala [hunk 3]
```diff
@@ -58,7 +59,7 @@ def toDouble(s: String, default:Double=Double.NaN) = {if (isNumeric(s)) s.toDoub
 
 
 val file_list = System.getenv("DNS_PATH")
-val feedback_file = "None"
+val feedback_file = System.getenv("LPATH") + "/dns_scores.csv"
 val duplication_factor = 100
```
Comment (NathanSegerlind) [inline dns_pre_lda.scala]: at present, this should be read from the System.getenv("DUPFACTOR")   per the code in 1.0.1 version of ml_ops.sh

in the near future, I'd like to change it to a command line argument,  but that's for another checkin, not today

File: dns_pre_lda.scala [hunk 4]
```diff
@@ -78,51 +79,97 @@ val l_top_domains = Source.fromFile("top-1m.csv").getLines.map(line => {
...
-val file_list = System.getenv("DPATH")    //This can be one file or many file pathes separated by commas
-val domain_table = System.getenv("DOMAIN_TABLE")
-val feedback_file = System.getenv("FPATH")
-val duplication_factor = System.getenv("DUPFACTOR").toInt
-val output_file = System.getenv("HPATH") + "/dns_word_counts"
-val output_file_for_lda = System.getenv("HPATH") + "/dns_lda_word_counts"
-*/
-
-//-----------------------------
+case class Feedback(frame_time: String,
```
Comment (NathanSegerlind) [inline dns_pre_lda.scala]: Python style variable names => camelCased Scala variable names

File: dns_pre_lda.scala [hunk 5]
```diff
@@ -78,51 +79,97 @@ val l_top_domains = Source.fromFile("top-1m.csv").getLines.map(line => {
...
         if (index > 1) {
-	    df = df.unionAll(sqlContext.parquetFile(file).filter("frame_len is not null and unix_tstamp is not null"))
-	}
+            df = df.unionAll(sqlContext.parquetFile(file).filter("frame_len is not null and unix_tstamp is not null"))
+        }
     }
     df =  df.select("frame_time","unix_tstamp","frame_len", "ip_dst", "dns_qry_name", "dns_qry_class", "dns_qry_type", "dns_qry_rcode")
     df_cols = df.columns
-    val tempRDD: org.apache.spark.rdd.RDD[String] = df.map(_.mkString(",")) 
+    val tempRDD: org.apache.spark.rdd.RDD[String] = df.map(_.mkString(","))
```
Comment (NathanSegerlind) [inline dns_pre_lda.scala]: it's not necessary to create the variable tempRDD here... if the last statement of the block were df.map(_.mkString(",")),  the same value would get returned

File: dns_pre_lda.scala [hunk 6]
```diff
@@ -78,51 +79,97 @@ val l_top_domains = Source.fromFile("top-1m.csv").getLines.map(line => {
...
+                                                                               row(DnsQryRcodeIndex),
+                                                                               row(DnsSevIndex).trim.toInt )).toDF()
+    val result = feedback_df.filter("dns_sev = 3").select("frame_time","unix_tstamp","frame_len", "ip_dst", "dns_qry_name", "dns_qry_class", "dns_qry_type", "dns_qry_rcode").map(_.mkString(","))
+    val toDuplicate : org.apache.spark.rdd.RDD[String] = result.flatMap(x => List.fill(duplication_factor)(x))
+    toDuplicate
+} else{
+        null
+}
 
 var multidata = {
```
Comment (NathanSegerlind) [inline dns_pre_lda.scala]: there are a lot of var declarations in this code... is this necessary? I bet some can be changed to val declarations

val declarations are simpler to debug and they can sometimes lead to performance increases for distributed code (less locking and synchronization)

File: dns_pre_lda.scala [hunk 7]
```diff
@@ -78,51 +79,97 @@ val l_top_domains = Source.fromFile("top-1m.csv").getLines.map(line => {
...
-    val tempRDD: org.apache.spark.rdd.RDD[String] = df.map(_.mkString(",")) 
+    val tempRDD: org.apache.spark.rdd.RDD[String] = df.map(_.mkString(","))
     tempRDD
 }
 
 var rawdata :org.apache.spark.rdd.RDD[String] = {
-    if (feedback_file == "None") { multidata
+    if (!scoredFileExists) { multidata
     }else {
         var data :org.apache.spark.rdd.RDD[String] = multidata
```
Comment (NathanSegerlind) [inline dns_pre_lda.scala]: three lines could be simplified to one:

multidata.union(falsePositives)

File: dns_pre_lda.scala [hunk 8]
```diff
@@ -58,7 +59,7 @@ def toDouble(s: String, default:Double=Double.NaN) = {if (isNumeric(s)) s.toDoub
 
 
 val file_list = System.getenv("DNS_PATH")
```
Comment (NathanSegerlind) [inline dns_pre_lda.scala:48]: python variable names => camelCased Scala variable names

File: dns_pre_lda.scala [hunk 9]
```diff
@@ -78,51 +79,97 @@ val l_top_domains = Source.fromFile("top-1m.csv").getLines.map(line => {
...
+                        frame_len: Int,
+                        ip_dst: String,
+                        dns_qry_name: String,
+                        dns_qry_class: String,
+                        dns_qry_type: String,
+                        dns_qry_rcode: String,
+                        dns_sev: Int)
+val scoredFileExists = new java.io.File(feedback_file).exists
+
+var falsepositives : org.apache.spark.rdd.RDD[String]  = if (scoredFileExists){
```
Comment (NathanSegerlind) [inline dns_pre_lda.scala]: falsePositives (camelCase)

**Comment by NathanSegerlind** [comment]
This looks pretty good except for some minor style issues.

My biggest concern is merging it into the jar-ified, librari-fied code that we have on my forked repo... might have to do a hand merge but the file is so small it shouldn't be hard.

File: dns_pre_lda.scala (hunk 1 above)
Comment (rabarona) [inline dns_pre_lda.scala:109]: Yea, I was trying to follow Scala naming conventions as per their documentation but I hear what you say. 
http://docs.scala-lang.org/style/naming-conventions.html
image

Comment (NathanSegerlind) [inline dns_pre_lda.scala:109]: well, okay then, I can't argue with that

we'll need to pull together a code style document for Scala/Spark code in the ML module...

some of my code violates that one  (I had been doing all caps for constants, like is done in C...)

File: dns_pre_lda.scala (hunk 3 above)
Comment (rabarona) [inline dns_pre_lda.scala]: Will change

File: dns_pre_lda.scala (hunk 4 above)
Comment (rabarona) [inline dns_pre_lda.scala]: will change

**Comment by rabarona** [comment]
@NathanSegerlind  I made the changes, need to test to make sure I didn't break anything.
Please take a look again and let me know if you find something else.

Thanks.

**Comment by NathanSegerlind** [comment]
looks good to me, just awaiting the final run...

**Comment by rabarona** [comment]
I was able to execute a final run an validate styling changes didn't break anything. 
Although I found this issue: https://github.com/Open-Network-Insight/open-network-insight/issues/72 
Fix: https://github.com/Open-Network-Insight/oni-ml/pull/25 - already included in this PR.

@NathanSegerlind @EverLoSa

**Pull request merged**